src/test/crimson/test_messenger.cc

```diff
@@ -3541,35 +3542,49 @@ int main(int argc, char** argv)
      "create a local crimson testpeer, or connect to a remote testpeer");
   return app.run(argc, argv, [&app] {
-    auto&& config = app.configuration();
-    verbose = config["verbose"].as<bool>();
-    auto rounds = config["rounds"].as<unsigned>();
-    auto keepalive_ratio = config["keepalive-ratio"].as<double>();
-    entity_addr_t v2_test_addr;
-    ceph_assert(v2_test_addr.parse(
+    std::vector<const char*> args;
+    std::string cluster;
+    std::string conf_file_list;
+    auto init_params = ceph_argparse_early_args(args,
+                                                CEPH_ENTITY_TYPE_CLIENT,
+                                                &cluster,
+                                                &conf_file_list);
+    return crimson::common::sharded_conf().start(init_params.name, cluster)
+    .then([conf_file_list] {
+      return local_conf().parse_config_files(conf_file_list);
+    }).then([&app] {
+      auto&& config = app.configuration();
+      verbose = config["verbose"].as<bool>();
+      auto rounds = config["rounds"].as<unsigned>();
+      auto keepalive_ratio = config["keepalive-ratio"].as<double>();
+      entity_addr_t v2_test_addr;
+      ceph_assert(v2_test_addr.parse(
           config["v2-test-addr"].as<std::string>().c_str(), nullptr));
-    entity_addr_t v2_testpeer_addr;
-    ceph_assert(v2_testpeer_addr.parse(
+      entity_addr_t v2_testpeer_addr;
+      ceph_assert(v2_testpeer_addr.parse(
           config["v2-testpeer-addr"].as<std::string>().c_str(), nullptr));
-    auto v2_testpeer_islocal = config["v2-testpeer-islocal"].as<bool>();
-    return test_echo(rounds, keepalive_ratio, false)
-    .then([rounds, keepalive_ratio] {
-      return test_echo(rounds, keepalive_ratio, true);
-    }).then([] {
-      return test_concurrent_dispatch(false);
-    }).then([] {
-      return test_concurrent_dispatch(true);
-    }).then([] {
-      return test_preemptive_shutdown(false);
-    }).then([] {
-      return test_preemptive_shutdown(true);
-    }).then([v2_test_addr, v2_testpeer_addr, v2_testpeer_islocal] {
-      return test_v2_protocol(v2_test_addr, v2_testpeer_addr, v2_testpeer_islocal);
+      auto v2_testpeer_islocal = config["v2-testpeer-islocal"].as<bool>();
+      return test_echo(rounds, keepalive_ratio, false)
+      .then([rounds, keepalive_ratio] {
+        return test_echo(rounds, keepalive_ratio, true);
+      }).then([] {
+        return test_concurrent_dispatch(false);
+      }).then([] {
+        return test_concurrent_dispatch(true);
+      }).then([] {
+        return test_preemptive_shutdown(false);
+      }).then([] {
+        return test_preemptive_shutdown(true);
+      }).then([v2_test_addr, v2_testpeer_addr, v2_testpeer_islocal] {
+        return test_v2_protocol(v2_test_addr, v2_testpeer_addr, v2_testpeer_islocal);
+      }).then([] {
+        std::cout << "All tests succeeded" << std::endl;
+        // Seastar has bugs to have events undispatched during shutdown,
```

Comment:
might be bug on our end? because we don't drain the events before shutdown the reactor?

Reply:
Please refer to https://github.com/ceph/ceph/pull/32632#discussion_r375041971

A related ASAN error occurred in my test environment, I think we cannot drain that event in our code as the future is discarded.

Reply:
@cyx1231st could you file an issue at https://github.com/scylladb/seastar/issues ?

Reply:
@tchaikov see https://github.com/scylladb/seastar/issues/738